addons/hr_holidays/models/hr_leave_type.py

```diff
@@ -320,6 +321,43 @@ def _search(self, domain, offset=0, limit=None, order=None, access_rights_uid=No
             return leaves._as_query()
         return super()._search(domain, offset, limit, order, access_rights_uid)
 
+    def _has_discrepancies(self, employees, target_date=None):
+        data = self.get_allocation_data(employees, target_date)
+        for employee_data in data.values():
+            for leave_type_data in employee_data:
+                if leave_type_data[1]['allows_negative']:
+                    if leave_type_data[1]['total_virtual_excess'] > leave_type_data[1]['max_allowed_negative']:
+                        return True
+                elif leave_type_data[1]['virtual_excess_data']:
+                    return True
+        return False
+
+    def write(self, values):
+        unchangeable_fields = {
+            'requires_allocation',
+            'allocation_type',
+            'request_unit',
+        }
+        change_unchangeable = bool(set(values.keys()).intersection(unchangeable_fields))
+        if values.get('allows_negative', True)\
+                and values.get('max_allowed_negative', float('inf')) >= max(self.mapped('max_allowed_negative') or [0])\
+                and not change_unchangeable:
+            return super().write(values)
+        leaves = self.env['hr.leave'].search([('holiday_status_id', 'in', self.ids)])
+        if change_unchangeable and (
+            self.env['hr.leave.allocation'].search([('holiday_status_id', 'in', self.ids)])
+            or leaves
+        ):
+            raise ValidationError(_(
+                "Fields modification is limited for leave types linked to at least one allocation or leave."
+            ))
+        res = super().write(values)
+        if self._has_discrepancies(leaves.employee_id):
+            raise ValidationError(_(
+                "The modification you're trying to do to negative amounts will bring some discrepancies to some employees."
+            ))
```

Comment:
Also this error message is a little hard to read to me. Does this mean that the changes being done result in some employees being further in negative than they are allowed to be?